module/state_synchronization/requester/oneshot_execution_data_requester_test.go

```diff
@@ -0,0 +1,238 @@
+		edCache := cache.NewExecutionDataCache(downloader, headers, seals, results, heroCache)
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		// Test each sealed block
+		for blockID, expectedED := range testData.executionDataByID {
+			block := testData.blocksByID[blockID]
+			result := testData.resultsByBlockID[blockID]
+
+			requester, err := NewOneshotExecutionDataRequester(
+				logger,
+				metricsCollector,
+				edCache,
+				result,
+				block.Header,
+				config,
+			)
+			require.NoError(suite.T(), err)
+			require.NotNil(suite.T(), requester)
+
+			execData, err := requester.RequestExecutionData(ctx)
+			require.NoError(suite.T(), err)
+
+			require.Equal(suite.T(), execData.BlockID, result.BlockID)
+			require.Equal(suite.T(), expectedED.BlockID, execData.BlockExecutionData.BlockID)
+			require.Equal(suite.T(), expectedED.ChunkExecutionDatas, execData.BlockExecutionData.ChunkExecutionDatas)
+		}
+	})
+}
+
+func (suite *OneshotExecutionDataRequesterSuite) TestRequestExecution_ERCacheReturnsError() {
+	logger := unittest.Logger()
+	metricsCollector := metrics.NewNoopCollector()
+	config := OneshotExecutionDataConfig{
+		FetchTimeout:    DefaultFetchTimeout,
+		MaxFetchTimeout: DefaultMaxFetchTimeout,
+		RetryDelay:      DefaultRetryDelay,
+		MaxRetryDelay:   DefaultMaxRetryDelay,
+	}
+
+	headers := storagemock.NewHeaders(suite.T())
+	seals := storagemock.NewSeals(suite.T())
+
+	// Create a block and execution result
+	block := unittest.BlockFixture()
+	executionResult := unittest.ExecutionResultFixture(unittest.WithBlock(&block))
+	results := storagemock.NewExecutionResults(suite.T())
+
+	suite.Run("blob not found error", func() {
+		// Mock downloader to return blob not found error first, then success
+		downloader := edmock.NewDownloader(suite.T())
+		expectedError := &execution_data.BlobNotFoundError{}
+		downloader.
+			On("Get", mock.Anything, executionResult.ExecutionDataID).
+			Return(nil, expectedError).
+			Once()
+
+		// eventually return execution data
+		blockEd := unittest.BlockExecutionDataFixture(unittest.WithBlockExecutionDataBlockID(block.ID()))
+		downloader.
+			On("Get", mock.Anything, executionResult.ExecutionDataID).
+			Return(blockEd, nil).
+			Once()
+
+		heroCache := herocache.NewBlockExecutionData(subscription.DefaultCacheSize, logger, metricsCollector)
+		edCache := cache.NewExecutionDataCache(downloader, headers, seals, results, heroCache)
+		requester, err := NewOneshotExecutionDataRequester(logger, metricsCollector, edCache, executionResult, block.Header, config)
+		require.NoError(suite.T(), err)
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		execData, err := requester.RequestExecutionData(ctx)
+		require.NoError(suite.T(), err)
+		require.NotNil(suite.T(), execData)
+		require.Equal(suite.T(), block.ID(), execData.BlockID)
+	})
+
+	suite.Run("execution data not found in storage", func() {
```

Comment:
this test should fail. `RequestExecutionData` should retry indefinitely when `execution_data.BlobNotFoundError` is returned from the downloader.